apps/src/templates/teacherDashboard/AssignmentSelector.jsx

```diff
@@ -172,17 +179,16 @@ export default class AssignmentSelector extends Component {
     return assignmentId(primaryAssignment.courseId, primaryAssignment.scriptId);
   }
 
-  setPrimary = (selectedAssignmentFamily, selectedVersion) => {
-    const versions = this.getVersions(selectedAssignmentFamily);
-    const recommendedVersion = versions.find(v => v.isRecommended);
-    selectedVersion = selectedVersion || recommendedVersion || versions[0];
-    const selectedPrimaryId = this.getSelectedPrimaryId(selectedAssignmentFamily, selectedVersion);
+  setPrimary = (selectedAssignmentFamily, selectedVersionYear) => {
+    const versions = this.getVersions(selectedAssignmentFamily, selectedVersionYear);
+    const selectedVersion = versions.find(v => v.isSelected);
+    selectedVersionYear = selectedVersion && selectedVersion.year;
+    const selectedPrimaryId = this.getSelectedPrimaryId(selectedAssignmentFamily, selectedVersionYear);
```

Comment:
The one bit of this code that tied my brain in knots is what we're doing with selectedVersionYear / isSelected through here.  I take it picking a selected version is necessarily a multi-step process with logic spread across a few methods?  I think the UI tests are mostly sufficient, but I wonder if some supplemental unit tests for this particular behavior would be good to have.

Reply:
Yes, this part is a bit perplexing. I've made a couple improvements, and I'll give a bit of additional info to help clear up any remaining confusion.

In the spirit of not duplicating state, the source of truth for the selected version year is the entry in this.state.version which has `isSelected` set. I stored it here because we need this set on the version when we pass it to the AssignmentVersionSelector (arguably not a great reason).

The selected version year is determined as follows:
1. the version year of the primary assignment, When the component is initialized in the `constructor`
2. the version selected by the user in the Version dropdown (calls setPrimary with `selectedVersionYear`)
3. when the user selects an Assignment Family (calls `setPrimary` without `selectedVersionYear`), the selected version year is set to:
  a. the year of the recommended version (if there is one), or
  b. the year of first version (if there are any), or
  c. undefined

I've made a couple small changes to clear this up. Any remaining confusion probably comes from a few of the following issues:
A. redundant state (`selectedAssignmentFamily` and the `isSelected` version in `versions` can be used to compute the primaryAssignmentId, and vice versa)
B. this.state.versions is an odd place to store the selected version year, and the reason for doing it (because a child component needs it) is somewhat backwards.

Possible improvements: 
X. store this.state.selectedVersionYear instead of storing isSelected on this.state.versions
Y. compute selectedPrimaryId from selectedAssignmentFamily and selectedVersionYear when needed instead of storing this.state.selectedPrimaryId
Z. stop storing this.state.versions. instead, use a "memoized selector" to compute them on the fly from this.state and this.props.assignments. this might require moving some of this.state into redux first.

The general pattern of these proposed improvements is to move the state closer to "what I received from the network" and "what is currently selected in the ui", rather than in boiled down abstract concepts like primary id and versions list.

Any recommendation on which if any of this to pursue, @islemaster?

Reply:
I'm actually fine with just the cleanup you've already done - getting rid of the re-assignment of `selectedVersionYear` makes the whole thing a lot easier to read.